drivers: counter: sam: Introduce sama7g5 pit64b

MCHP-MPU-Solutions-SHA opened the pull request:
Add counter driver for sama7g5 pit64b IP, use it as a standard zephyr counter device.
pit64 has only one compare register, by default pit64 counter device only have top_value channel, and have no alarm channel.

Didn't add support in samples/drivers/counter/alarm because pit64b counter device have no alarm channel by default.
Didn't add support in tests/drivers/counter/counter_basic_api because pit64b has no get_value() api, and get_value_64() should be used.

nandojve reviewed (changes requested):
Hi Xing,

You need to follow always the dependency order. In this case, you need to reorder the commits:

dts: arm: microchip: sam: add pit64b device to sama7g5
drivers: counter: add support for sama7g54 PIT64B

Then you need to proper extract and name those commits:

west: Add mchp sam pit64b dependency (when applicable)
dts: bindings: counter: Add mchp sam pit64b
soc: ? You need to add the early init because it is a pre-req to the driver.
drivers: counter: Introduce pit64b driver
boards: ? You need to consume the driver on a board
tests: ? You should enable the tests using that board (when applicable)

In the first commit, you create the binding and already can add the relevant nodes inside MPU dtsi.
In the drivers commit, you will put all the driver/kconfig/cmake work at once.
Finally you can consume the driver on a board.

This sequence allow each maintainer look in the topics that are relevant and do not break git bisect. If we see that something is missing in the hal, the west changes should be the first commit in the series using the reference `pull/<pr number>/head`.

BTW, we do not accept a driver that it is not build on a test or on a board.

MCHP-MPU-Solutions-SHA committed:
dts: arm: microchip: sam: add pit64b device to sama7g5

Add pit64b1 ~ pit64b5 counter devices to sama7g5

Signed-off-by: CHEN Xing <xing.chen@microchip.com>

MCHP-MPU-Solutions-SHA committed:
drivers: counter: add support for sama7g54 PIT64B

Add driver for sama7g54 PIT64B
Use PIT64B as a zephyr counter device

Signed-off-by: CHEN Xing <xing.chen@microchip.com>

MCHP-MPU-Solutions-SHA committed:
soc: microchip: sam: update mmu for sama7g5 pit64b

Update mmu region for mchp sam pit64b
Add gclk configuration for mchp sam pit64b

Signed-off-by: CHEN Xing <xing.chen@microchip.com>

MCHP-MPU-Solutions-SHA committed:
boards: microchip: sam: add and enable pit64b1

Add and enable pit64b1 counter device.

Signed-off-by: CHEN Xing <xing.chen@microchip.com>

MCHP-MPU-Solutions-SHA committed:
samples: drivers: counter: alarm: add definition for alarm flag

Counter devices can use this definition to specify alarm flags.

Signed-off-by: CHEN Xing <xing.chen@microchip.com>

MCHP-MPU-Solutions-SHA committed:
samples: drivers: counter: alarm: add support for pit64b1

Add support for pit64b1 counter device.

Signed-off-by: CHEN Xing <xing.chen@microchip.com>

MCHP-MPU-Solutions-SHA commented:
> Hi Xing,
> 
> You need to follow always the dependency order. In this case, you need to reorder the commits:
> 
> dts: arm: microchip: sam: add pit64b device to sama7g5 drivers: counter: add support for sama7g54 PIT64B
> 
> Then you need to proper extract and name those commits:
> 
> west: Add mchp sam pit64b dependency (when applicable) dts: bindings: counter: Add mchp sam pit64b soc: ? You need to add the early init because it is a pre-req to the driver. drivers: counter: Introduce pit64b driver boards: ? You need to consume the driver on a board tests: ? You should enable the tests using that board (when applicable)
> 
> In the first commit, you create the binding and already can add the relevant nodes inside MPU dtsi. In the drivers commit, you will put all the driver/kconfig/cmake work at once. Finally you can consume the driver on a board.
> 
> This sequence allow each maintainer look in the topics that are relevant and do not break git bisect. If we see that something is missing in the hal, the west changes should be the first commit in the series using the reference `pull/<pr number>/head`.
> 
> BTW, we do not accept a driver that it is not build on a test or on a board.

Hi Gerson,

Thank you so much for this detailed guide!

I have reorder these commits, please help to check.
Here is the updated commit list:
1) dts: arm: microchip: sam: add pit64b device to sama7g5
	dts/arm/microchip/sam/sama7g5.dtsi

2) drivers: counter: add support for sama7g54 PIT64B
	drivers/counter/Kconfig
	drivers/counter/Kconfig.mchp_sam_pit64b
	drivers/counter/counter_mchp_sam_pit64b.c

3) dts: bindings: counter: add mchp sam pit64b
	dts/bindings/counter/microchip,sam-pit64b-counter.yaml

4) soc: microchip: sam: update mmu for sama7g5 pit64b
	soc/microchip/sam/sama7g5/soc.c

5) drivers: counter: introduce mchp sam pit64b driver
	drivers/counter/CMakeLists.txt

6) boards: microchip: sam: add and enable pit64b1
	boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts

For this issue:
> BTW, we do not accept a driver that it is not build on a test or on a board.
Please check commit ff8d0eff026270cfbf55a61508b6e8cb576ce282 and daff86a132f8e4cf6b310296aa910b6afd418313.
Will you accept this solution? because PIT64B can only accept "COUNTER_ALARM_CFG_ABSOLUTE" alarm flag.
With these two commits pit64b counter device works well with zephyr/samples/drivers/counter/alarm.

Many thanks.
Xing.

nandojve commented on dts/bindings/counter/microchip,sam-pit64b-counter.yaml:
```diff
@@ -0,0 +1,42 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: Microchip SAM PIT64B Counter
```
This content should be merged with the 6889272722ed28665801db59930538924a5cf115

nandojve commented on drivers/counter/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_GECKO_STIMER        counter_gecko_st
 zephyr_library_sources_ifdef(CONFIG_COUNTER_IMX_EPIT            counter_imx_epit.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_ITE_IT51XXX         counter_ite_it51xxx.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_ITE_IT8XXX2         counter_ite_it8xxx2.c)
+zephyr_library_sources_ifdef(CONFIG_COUNTER_MCHP_SAM_PIT64B     counter_mchp_sam_pit64b.c)
```
This content should be merge with the 023be7f014b0df3b24196176f9a03ef6cda7e2f8, Remember that `CONFIG_COUNTER_MCHP_SAM_PIT64B` will be enabled only when devicetree `status == "okay"`. So, no damage will be made.

MCHP-MPU-Solutions-SHA replied:
Got it and understand, these two commits has been merged.

nandojve commented on dts/arm/microchip/sam/sama7g5.dtsi:
```diff
@@ -440,6 +440,51 @@
 		};
 
+		pit64b1: timer@e1804000 {
+			compatible = "microchip,sam-pit64b-counter";
```
this requires the binding from 3ff5f9562d24baaf62c8a4d361ae354f12ada64c

MCHP-MPU-Solutions-SHA replied:
Understand, thanks.

MCHP-MPU-Solutions-SHA commented on dts/bindings/counter/microchip,sam-pit64b-counter.yaml:
```diff
@@ -0,0 +1,42 @@
+
+compatible: "microchip,sam-pit64b-counter"
+
+include: [base.yaml, pinctrl-device.yaml]
```
Updated, many thanks for review.

MCHP-MPU-Solutions-SHA commented on drivers/counter/counter_mchp_sam_pit64b.c:
```diff
@@ -0,0 +1,438 @@
+DEVICE_DT_INST_DEFINE(n, sam_pit_init, NULL,				\
+		&counter_##n##_sam_data, &counter_##n##_sam_config,	\
+		POST_KERNEL, CONFIG_COUNTER_INIT_PRIORITY,		\
+		&sam_pit_driver_api);					\
```
Updated, many thanks for review.

nandojve replied:
```diff
index 8e6d24818bf..8a684af2cca 100644
--- a/drivers/counter/counter_mchp_sam_pit64b.c
+++ b/drivers/counter/counter_mchp_sam_pit64b.c
@@ -424,9 +424,10 @@ static const struct sam_pit_config counter_##n##_sam_config = {                    \
 static struct sam_pit_data counter_##n##_sam_data;                             \
                                                                                \
 DEVICE_DT_INST_DEFINE(n, sam_pit_init, NULL,                                   \
-       &counter_##n##_sam_data, &counter_##n##_sam_config,                     \
-       POST_KERNEL, CONFIG_COUNTER_INIT_PRIORITY,                              \
-       &sam_pit_driver_api);                                                   \
+                     &counter_##n##_sam_data,                                  \
+                     &counter_##n##_sam_config,                                \
+                     POST_KERNEL, CONFIG_COUNTER_INIT_PRIORITY,                \
+                     &sam_pit_driver_api);                                     \
                                                                                \
 static void counter_##n##_sam_config_func(const struct device *dev)            \
 {                                                                              \```

MCHP-MPU-Solutions-SHA replied:
Updated, many thanks for review.

nandojve commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts:
```diff
@@ -101,3 +101,9 @@
 &pit64b0 {
 	clock-frequency = <DT_FREQ_M(10)>;
 };
+
```
Missing the counter entry in https://github.com/zephyrproject-rtos/zephyr/blob/84de5b841713a0da97986e1edb3319edb354f904/boards/microchip/sam/sama7g54_ek/sama7g54_ek.yaml#L11
Add in alpha order

MCHP-MPU-Solutions-SHA replied:
Updated, many thanks for review.

MCHP-MPU-Solutions-SHA commented:
> Hi Xing,
> 
> I still don't see the driver in the list. Unfortunately, you need to force an entry here:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/5525941d839099309a44a0be868151fa894a5425/samples/drivers/counter/alarm/sample.yaml#L19
> 
> People are trying to reduce the number of board that compile the driver. This is something that we need to reevaluate in the project because, sometimes, we have different paths inside the driver.
> 
> So, only the boards in the list will build (no variations).
> 
> CC: @NhMch

Hi Gerson,
File zephyr/samples/drivers/counter/alarm/sample.yaml has been updated.
And verified with command:
west twister -p sama7g54_ek -s sample.drivers.counter.alarm

MCHP-MPU-Solutions-SHA commented:
> needs rebase

PR has been rebased.

pdgendt reviewed (commented):
No implicit boolean conversions in `if`-statements. https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#:~:text=85,N/A

pdgendt commented on drivers/counter/counter_mchp_sam_pit64b.c:
```diff
@@ -0,0 +1,439 @@
+{
+	const struct sam_pit_config *config = dev->config;
+	struct sam_pit_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&data->lock);
+
+	*ticks = (uint32_t)pit_counter_value(config->regs);
+
+	k_spin_unlock(&data->lock, key);
```
Alternative is to use `K_SPINLOCK`, which is a bit cleaner IMO (doesn't work with `return` statements in the block).


```c
	K_SPINLOCK(&data->lock) {
		*ticks = (uint32_t)pit_counter_value(config->regs);
	}
```

nandojve commented:
Xing,

I think you need to rebase due to CI issues.

MCHP-MPU-Solutions-SHA commented:
> Xing,
> 
> I think you need to rebase due to CI issues.

Gerson,

So, maybe this issue "Some checks were not successful" can be fixed by a rebase?
Actually I'm trying to fix it but can not find the root cause.

pdgendt commented:
> So, maybe this issue "Some checks were not successful" can be fixed by a rebase? Actually I'm trying to fix it but can not find the root cause.

Probably, a Python dependency was added on `main`

nandojve commented:
> > So, maybe this issue "Some checks were not successful" can be fixed by a rebase? Actually I'm trying to fix it but can not find the root cause.
> 
> Probably, a Python dependency was added on `main`

ahh, so if CI is broken then we need to wait the fix.
But we can make PR ready in the meantime : )

pdgendt commented:
> ahh, so if CI is broken then we need to wait the fix. But we can make PR ready in the meantime : )

A fix has already been merged to `main`, a rebase should get you going again.

MCHP-MPU-Solutions-SHA commented:
Many thanks for comments.
I will rebase this PR firstly.

MCHP-MPU-Solutions-SHA commented:
> ping

Sorry for later, this PR has been rebase.

nandojve reviewed (changes requested):
make sure CI is green.
small catch in devicetree

MCHP-MPU-Solutions-SHA commented:
> needs another rebase due to new merged content

PR has been rebase.

nandojve reviewed (approved):
Nice!
